File: app/app/src/main/java/dev/sheershbhatnagar/ai_assistant/core/network/SessionManager.kt
package dev.sheershbhatnagar.ai_assistant.core.network

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.*
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map

private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "user_session")

class SessionManager(private val context: Context) {
    companion object {
        val HOST_ADDRESS = stringPreferencesKey("host_address")
        val JWT_TOKEN = stringPreferencesKey("jwt_token")
        val USER_ID = stringPreferencesKey("user_id")
        val USER_EMAIL = stringPreferencesKey("user_email")
        val USER_FIRST_NAME = stringPreferencesKey("user_first_name")
        val USER_LAST_NAME = stringPreferencesKey("user_last_name")
        val DEFAULT_MODEL_ID = stringPreferencesKey("default_model_id")
        val IS_DARK_THEME = booleanPreferencesKey("is_dark_theme")
    }

    val hostAddress: Flow<String?> = context.dataStore.data.map { it[HOST_ADDRESS] }
    val jwtToken: Flow<String?> = context.dataStore.data.map { it[JWT_TOKEN] }
    val userId: Flow<String?> = context.dataStore.data.map { it[USER_ID] }
    val userEmail: Flow<String?> = context.dataStore.data.map { it[USER_EMAIL] }
    val userFirstName: Flow<String?> = context.dataStore.data.map { it[USER_FIRST_NAME] }
    val userLastName: Flow<String?> = context.dataStore.data.map { it[USER_LAST_NAME] }
    val defaultModelId: Flow<String?> = context.dataStore.data.map { it[DEFAULT_MODEL_ID] }
    val isDarkTheme: Flow<Boolean> = context.dataStore.data.map { it[IS_DARK_THEME] ?: true }

    suspend fun saveSession(host: String, token: String, userIdStr: String, email: String, firstName: String) {
        context.dataStore.edit { prefs ->
            prefs[HOST_ADDRESS] = host
            prefs[JWT_TOKEN] = token
            prefs[USER_ID] = userIdStr
            prefs[USER_EMAIL] = email
            prefs[USER_FIRST_NAME] = firstName
        }
    }

    suspend fun saveUserProfile(firstName: String, lastName: String?, email: String) {
        context.dataStore.edit { prefs ->
            prefs[USER_FIRST_NAME] = firstName
            if (lastName != null) {
                prefs[USER_LAST_NAME] = lastName
            } else {
                prefs.remove(USER_LAST_NAME)
            }
            prefs[USER_EMAIL] = email
        }
    }

    suspend fun saveThemeSetting(isDark: Boolean) {
        context.dataStore.edit { prefs ->
            prefs[IS_DARK_THEME] = isDark
        }
    }

    suspend fun saveHostAddress(host: String) {
        context.dataStore.edit { prefs ->
            prefs[HOST_ADDRESS] = host
        }
    }

    suspend fun saveDefaultModel(modelId: String) {
        context.dataStore.edit { prefs ->
            prefs[DEFAULT_MODEL_ID] = modelId
        }
    }

    suspend fun clearSession() {
        context.dataStore.edit { prefs ->
            prefs.remove(JWT_TOKEN)
            prefs.remove(USER_ID)
            prefs.remove(USER_EMAIL)
            prefs.remove(USER_FIRST_NAME)
            prefs.remove(USER_LAST_NAME)
            prefs.remove(DEFAULT_MODEL_ID)
        }
    }
}

File: app/lib/core/network/session_manager.dart
import 'package:shared_preferences/shared_preferences.dart';

class SessionManager {
  static const _keyHostAddress = 'host_address';
  static const _keyJwtToken = 'jwt_token';
  static const _keyUserId = 'user_id';
  static const _keyUserEmail = 'user_email';
  static const _keyUserFirstName = 'user_first_name';
  static const _keyUserLastName = 'user_last_name';
  static const _keyDefaultModelId = 'default_model_id';
  static const _keyIsDarkTheme = 'is_dark_theme';

  final SharedPreferences _prefs;

  SessionManager(this._prefs);

  String? get hostAddress => _prefs.getString(_keyHostAddress);
  String? get jwtToken => _prefs.getString(_keyJwtToken);
  String? get userId => _prefs.getString(_keyUserId);
  String? get userEmail => _prefs.getString(_keyUserEmail);
  String? get userFirstName => _prefs.getString(_keyUserFirstName);
  String? get userLastName => _prefs.getString(_keyUserLastName);
  String? get defaultModelId => _prefs.getString(_keyDefaultModelId);
  bool get isDarkTheme => _prefs.getBool(_keyIsDarkTheme) ?? true;

  Future<void> saveSession({
    required String host,
    required String token,
    required String userIdVal,
    required String email,
    required String firstName,
  }) async {
    await _prefs.setString(_keyHostAddress, host);
    await _prefs.setString(_keyJwtToken, token);
    await _prefs.setString(_keyUserId, userIdVal);
    await _prefs.setString(_keyUserEmail, email);
    await _prefs.setString(_keyUserFirstName, firstName);
  }

  Future<void> saveUserProfile({
    required String firstName,
    String? lastName,
    required String email,
  }) async {
    await _prefs.setString(_keyUserFirstName, firstName);
    if (lastName != null) {
      await _prefs.setString(_keyUserLastName, lastName);
    } else {
      await _prefs.remove(_keyUserLastName);
    }
    await _prefs.setString(_keyUserEmail, email);
  }

  Future<void> saveThemeSetting(bool isDark) async {
    await _prefs.setBool(_keyIsDarkTheme, isDark);
  }

  Future<void> saveHostAddress(String host) async {
    await _prefs.setString(_keyHostAddress, host);
  }

  Future<void> saveDefaultModel(String modelId) async {
    await _prefs.setString(_keyDefaultModelId, modelId);
  }

  Future<void> clearSession() async {
    await _prefs.remove(_keyJwtToken);
    await _prefs.remove(_keyUserId);
    await _prefs.remove(_keyUserEmail);
    await _prefs.remove(_keyUserFirstName);
    await _prefs.remove(_keyUserLastName);
    await _prefs.remove(_keyDefaultModelId);
  }
}
